Friday, October 16 2015, Minneapolis Fed

# Markov Perfect Equilibrium

QuantEcon https://python.quantecon.org/markov_perf.html and Chapter 7 (Section 7.6 and 7.7) of Ljungqvist and Sargent (2018).

Consider an economy with:
- 2 agents with utility $u_i(a_i, a_j, x)$ where $a_i$ denotes the action of $i$ and $x$ is the state variable.
- the future is discounted at rate $\beta$ so each agent, taking the strategy profile $a_{j}$ as given, solves:
\begin{align*}
\max_{(a_{it})_{t=0}^\infty} &\ \sum_{t=0}^\infty \beta^t \ u_i(a_i, a_j(h_t), x_t)\\
 & h_{t+1} = (h_t, (a_{it}, a_{jt}(h_t)))\\
 & x_{t+1} = f(a_{it}, a_{jt}(h_t), x_t)
\end{align*}
Note that each player takes into account the effect that her action has on the history and hence on the action of the other player but a player takes the strategy (not the sequence of actions) of the other player as given and unaffected by her choice of srategies.
- an action profile is Markovian if for every $i$, $a_i(x_t, h_{t-1}) = a_i(x_t, \tilde{h}_{t-1})$ for all histories $\tilde{h}_{t-1}$.
- we focus on Markov Perfect Equilibria. The Markovian best response of an agent can be formulated in recursive form:
\begin{align*}
v_i(x) = \max_{a_{i}} & \ u_i(a_i, a_j(x), x) + \beta v_i(x')\\
 & x' = f(a_i, a_j(x), x)
\end{align*}

A Markov Perfect Equilibrium is a pair of value functions and Markovian best responses $(v_i, a_i)_{i=1,2}$ such that $v_i$ solves the functional equation for agent $i$ and $a_i$ is the associated policy function.

The above problems are difficult to solve because they involve interrelated value functions iterations.
The FOC are, for all $x$:
\begin{equation*}
\frac{\partial u_i(a_i, a_j(x), x)}{\partial a_i} + \beta v_i'(f(a_1, a_2, x)) \frac{\partial f(a_i, a_j(x), x)}{\partial a_i} = 0
\end{equation*}
and by Benveniste Scheickman (NEED TO CHECK),
\begin{equation*}
v_i'(x') = \frac{\partial u_i(a_i, a_j(x'), x')}{\partial a_j} \frac{\partial a_j}{\partial x} + \frac{\partial u_i(a_i, a_j(x'), x')}{\partial x} 
\end{equation*}

Linear Quadratic case:
- notation, the choice variables are denoted by $u$ instead of $a$ 
- the objective is quadratic in $u_1, u_2, x$ with possibly an interaction term: $- \left(x^T R_i x + u_i^T Q_i u_i + u_j^T S_i u_j\right)$
- the transition function is linear $x_{t+1} = A x_t + B_1 u_{1t} + B_2 u_{2t}$.
- the value functions are quadratic $v_i = x^T P_i x$ and the policy functions are linear $a_i(x) = - F_i x$.
- the value function iteration for each $i$ takes the form of a Riccati equations, which now involve $F_j$. (Recall ,$i$ takes into account the effect of his action on the action that $j$ will take according to $F_j$ but takes $F_j$ as given.)
- the policy functions are the solution of a system (involving both $F_1$ and $F_2$) of linear equations.

Solution:
This problem can be solved by the method of undetermined coefficients:
- guess that $v_i$ is quadratic: $v_i(x) = x^T P_i x$ and the policy function is linear $u_i = -F_i x$.
- subsitute our guess for $v_i$ and $F_j$ in the functional equation for $i$.
The following Functional Equation obtains:
\begin{equation*}
x^T P_i x = \max_{u_i} - \bigg(x^T (R_i + F_j^T S_1 F_j) x + u_i^T Q_i u_i ) + \beta ((A - B_jF_j)x + B_i u_i)^T P_i ((A - B_jF_j)x + B_i u_i)\bigg)
\end{equation*}
- take FOC w.r.t $u_i$ to get the $P_i$-greedy (linear) policy function, given $F_j$.
Note that the above Functional equation has the same form as the one for an LQ dynamic programming problem except that $R$ is now $(R_i + F_j^T S_i F_j)$ and $A$ is now $(A - B_jF_j)$. So FOC give:
$u_i = -F_i x$ where 
\begin{equation*}
F_i = (Q_i + \beta B_i^T P_i B_i)^{-1} \beta (B_i^T P_i (A - B_jF_j))
\end{equation*}
- FOCs for $i=1,2$ give a system of linear equations, the solution to which is $F_1, F_2$
\begin{equation*}
\left\{
\begin{array}{cc}
F_1 &= (Q_1 +\beta B_1^T P_1 B_1)^{-1} \beta (B_1^T P_1 (A - B_2F_2))\\
F_2 &= (Q_2 +\beta B_2^T P_2 B_2)^{-1} \beta (B_2^T P_2 (A - B_1F_1))
\end{array}\right.
\end{equation*}
- substitute the linear policy function $F_i$ in the functional equation for $i$ to get rid of the max; in the LQ case, the resulting equation is an Algebraic Riccati equation. Note the Algebraic Riccati equation for $P_i$ still depends on $F_j$.
\begin{equation*}
P_i = (R_i + F_j^T S_i F_j) − (\beta B_i^T P_i (A - B_jF_j))^T(Q_i + \beta B_i^T P_i B_i)^{−1} (\beta B_i^T P_i (A - B_jF_j))+ \beta (A - B_jF_j)^T P_i (A - B_jF_j)
\end{equation*}

For the finite horizon case:
- first determine the final conditions $R_{i,f}$ such that payoff at $T$ is $x^T R_{i,f} x$ ($R_{i,f}$ is either given or solve the static simultaneous move game at the end)
- compute $F_{1,T-1}, F_{2,T-1}$ by solving the system of linear equations given by FOCs computed above, taking $R_{1,f}, R_{2,f}$ as the future value function $R_f := P_T$.
- compute $P_{i,T-1}$ by plugging $R_{i,f}$ as $P_T$ \emph{and} F_{j,T-1} in the Algebraic Riccati equation for $i$ (note, $F_{i,T-1}$ is already substituted, we did it to get rid of the max and obtain the Algebraic Riccati equation).
- continue by induction: first compute $F_{1,t-1}, F_{2,t-1}$ by solving the system of equations, using $P_{1,t}, P_{2,t}$. Then compute the value functions $P_{i,t-1}$ by plugging $P_{i,t}$ and $F_{j,t-1}$ in the Algebraic Riccati equation.

For the infinite horizon case:
-  take the limit of the finite horizon solution, hoping that it converges. The only difference is that the initial guess $P_i^0$ need not be $R_{i,f}$; it may be a good guess though. 

Note: the infinite horizon case is computed as the limit of the finite horion case but here, it is necessary to compute the policy function at every iteration to be able to compute the Riccati equation. An alternative could be to note compute the policy function every time to avoid solving the system of linear equations too much and compute the fixed point in $P_1, P_2$ for every step (this is very similar, or it may exactly be (TO CHECK) Howard's improvement algorithm); the new policy functions are computed only once the fixed point for the $P^{n-1}$-greedy policy functions is computed.


FALSE, THIS IS FOR LQ NON SINGLE AGENT
The solution is:
- policy function $u_i = -F_i x$ where $F_1,F_2$ solve the linear system of equations.
- value function $v_i(x) = x^T P_i x$ where $P_i$ solves its Algebraic Riccati Equation with the right policy function $F_j$
- the state evolves according to:
$x_{t+1} = (A - B_1 F_1 -  B_2 F_2) x_t$


To do: 
- Do the case with shocks. Does Certainty Equivalence still hold.
- Do the case with interaction terms between $u_1, u_2$ and $x$.

IMPORTANT: Note that the functional equation for each player has the same form as the one for an LQ dynamic programming problem except that
- $R$ is now $(R_i + F_j^T S_i F_j)$
- $A$ is now $(A - B_jF_j)$
- $Q$ is now $Q_i$
- $B$ is now $B_i$
- $P$ is now $P_i$
- $N = C = \epsilon = 0$
We can therefore use the functions that were used to solve LQ problem with the appropriate substitutions. The algorithm will now need, even for the infinite horizon case, to calculate $P^n$-greedy policy functions along the sequence to update $(R_i + F_j^T S_i F_j)$ and $(A - B_jF_j)$.

In [2]:
import matplotlib.pyplot as plt
%matplotlib inline
#from __future__ import division
import numpy as np
from numpy.linalg import inv
from numpy import dot

'''
Functions to compute Markov Perfect Equilibrium MPE of a Linear Quadratic environment
'''
def MPE_Fpolicy(beta, P_1, P_2, A, B_1, B_2, Q_1, Q_2):
    '''
    Returns the matrices F_1 and F_2 which characterize the P_1,P_2-greedy policy.
    It requires solving a system of FOCs, where each i takes F_j as given.
    Code that gives the result: F_1, F_2 = Markov_Fpolicy(...
    beta: discount factor
    P_i: value function v_i(x) = x^T P_i x
    A : effect of today's state on tomorrow's state
    B_i: effect of today's choice by i on tomorrow's state
    Q_i: direct effect of choice on payoff
    '''
    M_11 = Q_1 + beta * dot(B_1.T, dot(P_1, B_1))
    M_12 = beta * dot(B_1.T, dot(P_1, B_2))
    M_21 = beta * dot(B_2.T, dot(P_2, B_1))
    M_22 = Q_2 + beta * dot(B_2.T, dot(P_2, B_2)) 
    v_1 = beta * dot(B_1.T, dot(P_1, A))
    v_2 = beta * dot(B_2.T, dot(P_2, A))
    M = np.array([[M_11, M_12], [M_21, M_22]])
    M.shape = (k_1 + k_2, k_1 + k_2)
    v = np.array([v_1, v_2])
    v.shape = (k_1 + k_2, n)
    Sol = np.linalg.solve(M, v)
    return Sol[0:k_1], Sol[k_1: k_1+k_2]

def LQpolicy(F, x):
    '''
    Choice is u_i = - F_i x
    x : state variable(s)
    '''
    return - dot(F,x)

def LQvalue(x, P, beta, C):
    '''
    Value function: v_i(x) = x^T P_i x + d
    d accounts for the effect of risk on the value function
    '''
    d = (beta /(1 - beta)) * np.trace(dot(C.T, dot(P,C)))
    return dot(x.T, dot(P,x)) + d

def next_state_MPE(x, A, B_1, B_2, F_1, F_2, epsilon):
    '''
    Computes the next state according to x_{t+1} = A x_t + B_1 u_1 + B_2 u_2
    A: effect of today's state on tomorrow's state
    B_i: effect of today's choice by i on tomorrow's state
    F_i: Markovian best response is u_i = - F_i x
    '''
    return dot((A - dot(B_1,F_1) - dot(B_2,F_2)),x)

def update_Riccati_MPE_Fj(P, beta, A, B, R, Q, S, B_j, F_j):
    '''
    Updates the value function for a Linear Quadratic environment
    x^T P_i x = \max_{u_i} - \bigg(x^T (R_i + F_j^T S_1 F_j) x + u_i^T Q_i u_i ) + \beta ((A - B_jF_j)x + B_i u_i)^T P_i ((A - B_jF_j)x + B_i u_i)\bigg)
    P: inital value function
    A: effect of today's state on tomorrow's state
    B: effect of today's choice by i on tomorrow's state
    R: effect of state on payoff
    Q: direct effect of choice on payoff
    S: direct effect of other's choice on payoff (externality)
    B_j: effect of other's choice on tomorrow's state
    F_j: choice function u_j = - F_j x of the other, taken as given.
    '''
    New_R = R + dot(F_j.T, dot(S, F_j))
    New_A = A - dot(B_j, F_j)
    New_Z = (beta * dot(B.T, dot(P, New_A)))
    W = inv(Q + beta * dot(B.T, dot(P,B)))
    return New_R - dot(New_Z.T, dot(W, New_Z)) + beta * dot(New_A.T, dot(P, New_A))

def Riccati_MPE_F_j(init_P, beta, A, B, R, Q, S, B_j, F_j):
    #this could be useful to do something similar to Howard's improvement algorithm
    #the system of simultaneous equations would be solved less often and more Riccati iterations would be done.
    #should include a maximum number of iterations.
    dist = 1
    while dist > 10e-3:
        next_P = update_Riccati_MPE(init_P, beta, A, B, R, Q, S, B_j, F_j)
        dist = np.linalg.norm(init_P - next_P)
        init_P = next_P
    return next_P

In [7]:
def LQMarkovPerfect(max_iter, tol, init_P_1, init_P_2, beta, A, B_1, B_2, R_1, R_2, Q_1, Q_2, S_1, S_2):
    '''
    This is the main loop to compute Markov Perfect Equilibrium in a Linear Quadratic environment:
        - F_1, F_2 for Markovian Perfect Equilibrium functions u_i = - F_i x_i
        - P_1, P_2 for value functions of the Markov Perfect Equilibrium, v_i(x) = x^T P_i x
    LQ environment: x^T P_i x = \max_{u_i} - \bigg(x^T (R_i + F_j^T S_1 F_j) x + u_i^T Q_i u_i ) + \beta ((A - B_jF_j)x + B_i u_i)^T P_i ((A - B_jF_j)x + B_i u_i)\bigg)
    max_iter: stop if this number of iterations is reached
    tol: tolerance below which the sequence is near enough convergence
    init_P_i: initial guess of value function
    beta: discount factor
    A : effect of today's state on tomorrow's state
    B_i: effect of today's choice by i on tomorrow's state
    R_i: effect of state on payoff
    Q_i: direct effect of choice on payoff
    S_i: direct effect of other's choice on payoff (externality)
    '''
    for it in range(max_iter):
        # compute the init_P-greedy policy functions which will be fed into this round of value function iteration
        F_1, F_2 = MPE_Fpolicy(beta, init_P_1, init_P_2, A, B_1, B_2, Q_1, Q_2)
        F_1.shape = (k_1, n)
        F_2.shape = (k_2, n)
        # compute the next element of the sequence of value function on the way to convergence
        P_1 = update_Riccati_MPE_Fj(init_P_1, beta, A, B_1, R_1, Q_1, S_1, B_2, F_2)
        P_2 = update_Riccati_MPE_Fj(init_P_2, beta, A, B_2, R_2, Q_2, S_2, B_1, F_1)        
        #distance between the last two elements in the sequence
        dist_1 = np.linalg.norm(init_P_1 - P_1)
        dist_2 = np.linalg.norm(init_P_2 - P_2)
        #update and loop back unless the Cauchy sequence has converged
        init_P_1 = P_1
        init_P_2 = P_2
        if max(dist_1, dist_2) < tol:
            print(it, 'number of iterations')
            break
            return F_1, F_2, P_1, P_2, 'F_1, F_2, P_1, P_2'
    else:
        print('not converged')
    #compute the policy functions associated with the last value functions in from the loop.
    F_1, F_2 = MPE_Fpolicy(beta, P_1, P_2, A, B_1, B_2, Q_1, Q_2)
    return F_1, F_2, P_1, P_2

In [9]:
'''
Environment
'''
# Parameters
gamma = 12
a_0 = 10
a_1 = 2
beta = 0.96
#c = 2 
#rho = 0.9
#sigma = 0.15

# Elements of the LQ environment
n = 3 #number of state variables
k_1 = 1 #number of control variables for player 1
k_2 = 1 #number of control variables for player 2
R_1 = np.array([[0, -0.5 * a_0, 0],
                [-0.5 * a_0, a_1, 0.5 * a_1],
                [0, 0.5 * a_1, 0]])
R_2 = np.array([[0, 0, -0.5 * a_0],
                [0, 0, 0.5 * a_1],
                [-0.5 * a_0, 0.5 * a_1, a_1]])
Q_1 = np.array([gamma])
Q_2 = np.array([gamma])
S_1 = np.array([[0]])
S_2 = np.array([[0]])
A = np.array([[1, 0, 0],
              [0, 1, 0],
              [0, 0, 1]])
B_1 = np.array([[0], [1], [0]])
B_2 = np.array([[0], [0], [1]])

 
'''
Main loop
'''
max_iter = 1000 # maximum number of iterations
tol  = 10e-9 # tolerance level below which convergence is close enough
init_P_1 = np.eye(3) # initial guess of value function
init_P_2 = np.eye(3)

F_1, F_2, P_1, P_2 = LQMarkovPerfect(max_iter, tol, init_P_1, init_P_2, beta, A, B_1, B_2, R_1, R_2, Q_1, Q_2, S_1, S_2)
print(F_1), 'F_1'
print(F_2), 'F_2'

492 number of iterations
[[-0.66846613  0.29512482  0.07584666]]
[[-0.66846613  0.07584666  0.29512482]]


(None, 'F_2')

The above solution coincides with the code from QuantEcon https://python.quantecon.org/markov_perf.html#Background
\begin{align*}
F1 &= [[-0.66846615, \quad   0.29512482, \quad   0.07584666]]\\
F2 &= [[-0.66846615, \quad   0.07584666, \quad   0.29512482]]
\end{align*}